src/vm/threadsuspend.cpp

```diff
@@ -6191,9 +6191,84 @@ void Thread::UnhijackThread()
         // it.
 //       _ASSERTE(*m_ppvHJRetAddrPtr == OnHijackTripThread);
 
-        STRESS_LOG2(LF_SYNC, LL_INFO100, "Unhijacking return address 0x%p for thread %p\n", m_pvHJRetAddr, this);
+        bool unhijacked = false;
+
         // restore the return address and clear the flag
-        *m_ppvHJRetAddrPtr = m_pvHJRetAddr;
+#if defined(_TARGET_ARM_) || defined(_TARGET_ARM64_)
+        // On ARM architectures, the return address is sometimes not stored on the stack, but it is only in the LR register.
+        // In a corner case when a function tail calls another one, there is a small window when even the hijacked address is 
+        // temporarily not on the stack (after the LR is poped from the stack in the epilog of the caller and before it is
+        // stored back to the stack in the the prolog of the callee).
```

Comment:
Is it guaranteed that the LR is stored in the same place in the prolog of the callee?

Reply:
It is not, that's why I'm using the context pointer restored by the Unwind to determine the location.

Reply:
If the LR is not stored in the same place, I think it means that `m_pvHJRetAddr` points to garbage after the callee start running and unhijacking will corrupt the stack.

I think that the unwind to find the right LR location would have to be done always here, even for the same thread case.

Reply:
The `m_pvHJRetAddr` is the return address (function address) itself, not a pointer to its location on the stack (the `m_ppvHJRetAddrPtr` is the location)

Reply:
I meant to say in `m_ppvHJRetAddrPtr` in my comment - updated.

Reply:
This pointer is used only in case it is not in the current function's frame. If it is inside of the current function's frame, the LR context pointer that we get from the unwinding is used instead to make sure that what you've mentioned doesn't happen.

Reply:
Let's say we have a methods like this:

```
void A(int x, int y, int z)
{
    ... do stuff ...
    bar(x);
}

void B(int x)
{
    LargeStruct myStruct;
    ...
}
```

The JIT decides to tailcall B from A. So the code may look like this:

```
A:
    push {r0-r2}
    push {r11, lr}

    ... do stuff ... <- We hijack here. lr stored in the stack frame is hijack helper.
    
    pop {r11, lr}
    br B <- lr is hijack helper. Your fix handles unhijacking here fine. 
```

```
B:
   push {r0}
   push {r11, lr} <- lr stored is in 
   sub sp, 0x100 <- m_ppvHJRetAddrPtr points somewhere into this block

   ... do stuff ... <- What happens when we try to unhijack here?
```

Reply:
We don't use the m_ppvHJRetAddrPtr in this case, but use the LR context pointer we get from the unwind as I've said before. We detect this case by checking that the m_ppvHJRetAddrPtr is <= SP that we get by the unwind, which means that the m_ppvHJRetAddrPtr is in the space of the current frame and we cannot rely on it.

Reply:
`do stuff` can call arbitrarily deep. A single unwind is not guaranteed to get you above `m_ppvHJRetAddrPtr` so that `m_ppvHJRetAddrPtr <= callerContext.Sp` condition is satisfied.

Reply:
Also, `do stuff` can throw and catch exceptions. I do not think `GetThread()->UnhijackThread()` in `UnwindManagedExceptionPass1` will work.

Reply:
Ok, you're right, that's a hole in my thinking. I have not considered the case when the tail call causes change in the LR location somewhere up the call stack. Then you are right that it would not work and it also influences the unhijacking of current thread.
So it looks like there is no way to detect that a tail call caused this by inspecting stuff. Even if we tried to unwind multiple frames until we reach a frame that contains the address stored in m_ppvHJRetAddrPtr, it would not work - the original function might have had a huge frame and the function it has tail called just a small one, so the current m_ppvHJRetAddrPtr might be several frames away from the frame of the function that contains the new location that we need to modify.
I'll close this PR and start working on some other way.